Add notice for Square SOR and inventory sync in bulk edit screen

m-muhsin opened the pull request:
<img width="766" height="219" alt="image" src="https://github.com/user-attachments/assets/379df1f4-322c-4c1f-8c87-159cf4f13209" />

### All Submissions:

<!-- Mark completed items with an [x] -->
* [x] Does your code follow the [WooCommerce Sniffs](https://github.com/woocommerce/woocommerce-sniffs/) variant of WordPress coding standards?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?
* [ ] Will this change require new documentation or changes to existing documentation?

---

### Changes proposed in this Pull Request:
<!-- Describe the changes made to this Pull Request and the reason for such changes. -->
Adds the above notice.

Closes https://linear.app/a8c/issue/SQUARE-204/bulk-stock-changes-in-wc-dont-sync-to-square.

### Steps to test the changes in this Pull Request:
<!-- Describe the steps to replicate the issue and confirm the fix -->
<!-- Try to include as many details as possible. -->

1. Set up Square with WooCommerce as the System of Record (SoR)
2. Ensure "Sync inventory" is enabled in WooCommerce > Settings > Square
3. Create or select 2+ products that are synced with Square and have stock management enabled (e.g., Product A with qty 10, Product B with qty 11)
4. In Products list, select the products and click "Edit" under Bulk actions
5. In the bulk edit panel, notice the above screenshot.

### Changelog entry
<!-- 
Each line should start with change type prefix`(Add|Fix|Dev) - `.
If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.
Add the `changelog: none` label if no changelog entry is needed.
-->

> Add - Show a notice for Square SOR and inventory sync in bulk edit screen.

iamdharmesh commented on includes/Handlers/Products.php:
```diff
@@ -577,6 +583,117 @@ public function set_synced_with_square( $product ) {
+	 *
+	 * This hooks into the stock change events to capture previous values and sync to Square.
+	 *
+	 * @since 4.9.0
```
```suggestion
	 * @since x.x.x
```

We are using x.x.x version for the PR and it will get replace to actual version during the release process. Same applies to other places as well.

iamdharmesh reviewed (changes requested):
Thanks for the PR @m-muhsin.

While doing quick check, I noticed that if we update stock of product with manage stock disabled using bulk edit, it become **out of stock** after some time. 

This is due to how Inventory Sync work ([Docs](https://woocommerce.com/document/woocommerce-square/sync-settings/sync-product-data-from-square-to-woocommerce/#stock-update-process)). Inventory is always pulled from Square to account for stock updates on the Square side (e.g., POS transactions), even when WooCommerce is set as SOR. For Woo SOR, Inventory only syncs from Woo to Square during the initial sync or when stock quantity is updated on the product edit page by clicking "Fetch stock from Square."

Currently, we don't have real time updates for the stock updates at Square side. So, we are sync it every 15 mins. But there is always chance that stock on WooCommerce is not up to date. So, we suggests merchant to update stock from the product edit page by clicking "Fetch stock from Square." or directly update on Square itself (Regardless of SOR).

For adding support for the stock changes using the Bulk edit we need to keep these things in mind and need to update implementation accordingly. Happy to discuss this over call if needed. Thank you.

m-muhsin commented:
Hi @iamdharmesh,

> While doing quick check, I noticed that if we update stock of product with manage stock disabled using bulk edit, it become out of stock after some time.

I could replicate the issue you mentioned.

> This is due to how Inventory Sync work ([Docs](https://woocommerce.com/document/woocommerce-square/sync-settings/sync-product-data-from-square-to-woocommerce/#stock-update-process)). Inventory is always pulled from Square to account for stock updates on the Square side (e.g., POS transactions), even when WooCommerce is set as SOR. For Woo SOR, Inventory only syncs from Woo to Square during the initial sync or when stock quantity is updated on the product edit page by clicking "Fetch stock from Square."

> Currently, we don't have real time updates for the stock updates at Square side. So, we are sync it every 15 mins. But there is always chance that stock on WooCommerce is not up to date. So, we suggests merchant to update stock from the product edit page by clicking "Fetch stock from Square." or directly update on Square itself (Regardless of SOR).

> For adding support for the stock changes using the Bulk edit we need to keep these things in mind and need to update implementation accordingly. Happy to discuss this over call if needed. Thank you.

Agree on the above.

Also, there seems to be an issue here.

Please refer to this video:

https://github.com/user-attachments/assets/de5756a1-3b30-4ab8-b980-ce7327510fee

Individual stock updating is not being allowed. However, bulk updating of stock seems to work fine.

Perhaps we need to think about a different strategy here?

iamdharmesh commented:
Hi @m-muhsin,

> Individual stock updating is not being allowed. However, bulk updating of stock seems to work fine.

Yes, bulk stock updates work with this PR in some cases, but as I mentioned [earlier](https://github.com/woocommerce/woocommerce-square/pull/439#pullrequestreview-3577596752), it still has issues. We need to keep Square as the source of truth for inventory updates regardless of the SOR until real-time stock updates from Square -> WooCommerce are implemented via webhooks.

For now, I think the best next step would be to add a note near the stock field in bulk edit to inform users that inventory updates will not work for Square sync-enabled products, as the values are ultimately overwritten with data from Square. For reference, we already show a similar notice on the product import screen, and we could follow a similar approach here.

<img width="1019" height="309" alt="image" src="https://github.com/user-attachments/assets/4d390034-a1cf-42c4-aadc-c57d85514156" />

@dkotter @faisal-alvi tagging you here to check what do you think on this or if we have any better way to solve this. Thank you.

m-muhsin commented:
Thanks, @iamdharmesh. I will wait on the response from Darin / Faisal and take any next steps on this one.

dkotter commented:
> For now, I think the best next step would be to add a note near the stock field in bulk edit to inform users that inventory updates will not work for Square sync-enabled products, as the values are ultimately overwritten with data from Square. For reference, we already show a similar notice on the product import screen, and we could follow a similar approach here

Yeah, I think this is fine for now, noting it doesn't actually fix anything but at least it makes it more clear for the merchant

Commit:
Add notice for Square SOR and inventory sync in bulk edit screen

Implement a MutationObserver to display a warning message in the bulk edit form when Square SOR and inventory sync are enabled. The notice informs users that inventory updates will not apply to Square-synced products and will be overwritten during the next sync.

m-muhsin commented:
I have now updated the PR with the notice as shown in the screenshot on the PR description. Thanks, @iamdharmesh @dkotter

iamdharmesh commented on src/js/admin/wc-square-admin-products.js:
```diff
@@ -31,6 +31,39 @@ jQuery( document ).ready( ( $ ) => {
 	// products quick edit screen.
 	if ( 'edit-product' === pagenow ) {
+		// Add notice near stock field in bulk edit when Square SOR and inventory sync is enabled.
+		if ( wc_square_admin_products.is_square_sor && wc_square_admin_products.is_inventory_sync_enabled ) {
```
```suggestion
		if ( wc_square_admin_products.is_inventory_sync_enabled ) {
```
As discussed earlier, Square is source of truth for inventory updates regardless of the SOR.

m-muhsin committed:
Set Square as source of truth for inventory regardless of SoR.

Co-authored-by: Dharmesh Patel <dspatel44@gmail.com>

qasumitbagthariya reviewed (approved):
### QA Update  ✅ 

-------------

I have verified this PR in the `fix/sqr-204` branch, which has been fixed and is functioning as intended.

I tested the following on this branch:

- Bulk edit option works as expected, and the bulk edit panel displays correctly as shown in the screenshot.
- No issues observed during bulk editing.

<img width="1248" height="712" alt="image" src="https://github.com/user-attachments/assets/fea07c9d-cb4c-41f0-b14a-949485316028" />


**Testing Environment** 

<Details>

- WordPress: 6.9
- Theme: Twenty Twenty-Five 1.4
- WooCommerce - 10.4.0
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2 
- Branch: fix/sqr-204

</Details>


Steps to Test- As mentioned in the PR description.
Test Results - It is working as expected.
Functional Demo / Screencast - 
Special Notes -  Ready for UAT
Testing Document status:
Cases related to this Issue/PR are added to the Critical Flow Wiki pages: 

- [x] Yes
- [ ] Not Required/Applicable for this PR
------------------------------------------------

qasumitbagthariya commented:
Regression / Smoke Test Report  ✅ 
-------

Tested with Archive File created via "php woorelease.phar build repo_URL" (Composer version 2.8.4, npm version 10.8.2, node version 20.18.1)

Status- Working expected with Plugin Archive/Zip file same as fix specific branch.

**Testing Environment** 

<Details>

- WordPress: 6.9
- Theme: Storefront 4.6.2
- Theme: Twenty Twenty-Five 1.4
- WooCommerce - 10.4.3
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2 
- Branch: smoke-testing

</Details>

Next Step- Ready to Merge 🚀
